### PR TITLE
[v8r0] Treat X509 data as str rather than bytes and add DISET+JEncode CI

### PR DESCRIPTION
**Pull request by chrisburr**
(This explanation is rather rushed but it's all I have time for right now, I can discuss it further later if needed.)

The issue is that DEncode doesn't have a good distinction between `bytes` and `str` so when moving to Python 3 I made it treat everything as strings. In the handful of cases that bytes are truly needed I used the [surrogate escapes](https://www.python.org/dev/peps/pep-0383/). (This is a horrible hack but keeps DEncode working the way it always has.) With `JEncode` this isn't an option as the actual wire format is now defined to be text so it can't pass around the invalid unicode sequences.

The biggest break is that we intentionally decided to treat PEM-encoded x509 as `bytes`. I'd forgotten that we sent them over the network therefore `DEncode` could quietly comvert them to strings (because the PEM encoding is base64+ASCII based). This PR bites the bullet and moves the PEM certificate handling to use strings, thankfully the test coverage is decent so it should be mostly working.

This PR also fixes a bunch of other cases where bytes were accidentally used and adds CI that uses JEncode so we can avoid future breaks.

BEGINRELEASENOTES

*Core
CHANGE: Treat X509 data as str instead of bytes

*Framework
FIX: UserProfileManager usage with JEncode

*WorkloadManagement
FIX: JobDB usage with JEncode

*Production
FIX: ProductionManager use with JEncode

ENDRELEASENOTES

### CODE REVIEW

**Comment by chrisburr** [comment]
Thanks for the reminder :) I've updated the PR description.

File: .github/workflows/integration.yml [hunk 1]
```diff
@@ -37,6 +37,8 @@ jobs:
             ARGS: MYSQL_VER=mariadb:10.6.3 ES_VER=opensearchproject/opensearch:1.0.0
           - TEST_NAME: "HTTPS"
             ARGS: TEST_HTTPS=Yes
+          - TEST_NAME: "Force JEncode"
+            ARGS: DIRAC_USE_JSON_ENCODE=Yes DIRAC_USE_JSON_DECODE=Yes
```
Comment (fstagni) [inline .github/workflows/integration.yml:41]: ...just to say that this was the way to go! I don't know why we didn't think to do this before!

**Review by fstagni — APPROVED** [review]
This looks to me like the correct approach and I would give it a go in tomorrow's hackathon

File: src/DIRAC/FrameworkSystem/Client/UserProfileClient.py [hunk 2]
```diff
@@ -17,7 +17,7 @@ def __getRPCClient(self):
 
     def storeVar(self, varName, data, perms={}):
         try:
-            stub = DEncode.encode(data)
+            stub = DEncode.encode(data).decode()
```
Comment (TaykYoku) [inline src/DIRAC/FrameworkSystem/Client/UserProfileClient.py:20]: this class I think better inherit `Client`

**Review by TaykYoku — COMMENTED** [review]
Overall, it looks good to me

**Comment by TaykYoku** [comment]
I'm a little worried that such a change could affect extensions

File: src/DIRAC/FrameworkSystem/DB/ProxyDB.py [hunk 3]
```diff
@@ -427,7 +427,7 @@ def __storeProxy(self, userDN, chain, proxyProvider=None):
         sqlInsert = True
         if data:
             sqlInsert = False
-            pem = data[0][1]
+            pem = data[0][1].decode("ascii")
```
[the remark this answers is not in the document]
Comment (chrisburr) [reply src/DIRAC/FrameworkSystem/DB/ProxyDB.py:430]: Not really but I think it's safer to make clear that this should be decoded (or better yet, changed to `TEXT` in the database).

File: tests/Integration/Framework/Test_UserProfileClient.py [hunk 4]
```diff
@@ -41,7 +41,7 @@ def test_storeRetrieveDelete():
 
     retVal = up.listAvailableVars()
     assert retVal["OK"], retVal
-    assert retVal["Value"] == (("adminusername", currentGroup, userVO or "undefined", key),)
+    assert retVal["Value"] == [["adminusername", currentGroup, userVO or "undefined", key]]
```
[the remark this answers is not in the document]
Comment (chrisburr) [reply tests/Integration/Framework/Test_UserProfileClient.py:44]: I fixed this in the UserProfileDB to be consistent.

(Search for `# Convert returned tuples to lists to appease JEncode`)

[the remark this answers is not in the document]
Comment (TaykYoku) [reply tests/Integration/Framework/Test_UserProfileClient.py:44]: Ah, sorry, seems I missed that

File: src/DIRAC/FrameworkSystem/Client/UserProfileClient.py (hunk 2 above)
Comment (chrisburr) [reply src/DIRAC/FrameworkSystem/Client/UserProfileClient.py:20]: I don't understand what you mean here?

Comment (TaykYoku) [reply src/DIRAC/FrameworkSystem/Client/UserProfileClient.py:20]: I mean that `DEncode` or `JEncode` encoding should be inside `Client` logic, IIUC

Comment (chrisburr) [reply src/DIRAC/FrameworkSystem/Client/UserProfileClient.py:20]: The `UserProfileClient` stores `DEncode`-ed objects in the database. Maybe that should be changed but definitely not in this PR.

**Comment by chrisburr** [comment]
> I'm a little worried that such a change could affect extensions

How exactly? Are there any active extensions to the proxy manager?

**Comment by TaykYoku** [comment]
This for example https://github.com/DIRACGrid/WebAppDIRAC/pull/630

**Review by TaykYoku — APPROVED** [review]

**Pull request merged**